### DISCUSSION TITLE
Style components globally with Tailwind css classNames

### DISCUSSION THREAD

**Discussion by frec-hansen** [mantinedev/mantine#912] [Q&A] [answered]
Hi all, thanks for the great work! I was wondering if there is anyway to style component globally with Tailwind.
Let's say I want my `TextInput` to look something like:
```javascript
<TextInput
  required
  classNames={{
    root: "col-span-5",
    input: "border-gray-500 rounded-none border",
  }}
/>
```

Currently in order to style component, I can either:
- use `MantineProvider` and write all the style (which I don't want to since it comes with Tailwind)
```javascript
<MantineProvider
  styles={{
    TextInput: (theme) => ({
      input: { borderWidth: 1, borderColor: '#f3f3f3', borderRadius: 0 }
    })
  }}
>
```
- wrap the component up with my own: which defeats the purpose of having global style
```javascript
function MyInput(props) {
  return (
    <TextInput
      required
      classNames={{
        root: "col-span-5",
        input: "border-gray-500 rounded-none border",
      }}
    />
  )
}
```
None of the solutions above is ideal. Is it possible to do something like:
```javascript
<MantineProvider
  classNames={{
    TextInput: (theme) => ({
      input: `border-gray-500 rounded-none border`
    })
  }}
>
```
I know there is a macro [Tailwind Macro](https://github.com/ben-rogerson/twin.macro) that works well with `emotion` to convert from className to css. But I am not familiar with `emotion` at all and I don't know how it is being used under the hood with Mantine. I had no luck trying it out, could you take a look to see if this can be possible?

**Comment by rtivital** [comment, marked as answer]
That is possible, I'll add this feature in one of the future releases

**Reply by frec-hansen** [reply]
Also, is it possible to specify everything to be `unstyled` globally?

**Reply by rtivital** [reply]
Most of the components do not support the unstyled variant. You will be able to set default props for all components on MantineProvider when version 4.0 is out.

**Reply by frec-hansen** [reply]
Sounds good, can't wait for 4.0

**Reply by skelawsky** [reply]
```
<MantineProvider
  classNames={{
    TextInput: (theme) => ({
      input: `border-gray-500 rounded-none border`
   })
 }}>
```


Does this styling solution work in v4.0.9?
I'd really like to use tailwind with mantine. Global styling for components using classes would be a great solution.

**Reply by minhluuquang** [reply]
@skelawsky yes it is. You can take a look at the release here https://mantine.dev/changelog/4-1-0/

**Comment by bencehusi** [comment]
Do note everyone that the classNames prop has moved into the theme, see: https://mantine.dev/changelog/5-0-0/#mantineprovider-changes

It took some digging for me!

So it's now: 

```
<MantineProvider
      withGlobalStyles
      withNormalizeCSS
      theme={{
        colors: mantineColorsObjectResult(),
        components: {
          Button: {
            classNames: {
              root: 'bg-green-600',
            }
          },
          TextInput: {
            classNames: {
              input: `border-yellow-800 rounded-none border`
            }
          }
        }
      }}
    >
      <Component {...pageProps} />
    </MantineProvider>
```

However, there's one thing I can't figure out: 

How do I style component variants with classNames?

In my case, this className on the button will override all variants (outlined, etc...). 

How can I use CSS classNames for styling variants?
